File: Vb/ImageProcessing/GrayScale.vb
Imports System.Threading
Imports System.Runtime.InteropServices.Marshal

Namespace ImageProcessing
    Public Class GrayScale : Inherits ComImgProc
        Public Sub New(_bitmap As BitmapImage)
            MyBase.New(_bitmap)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
        End Sub

        Public Overrides Function GoImgProc(_token As CancellationToken) As Boolean
            Dim bRst As Boolean = True

            Dim nWidthSize As Integer = Bitmap.Width
            Dim nHeightSize As Integer = Bitmap.Height

            WriteableBitmap = New WriteableBitmap(Bitmap)
            WriteableBitmap.Lock()

            Dim nIdxWidth As Integer
            Dim nIdxHeight As Integer

            For nIdxHeight = 0 To nHeightSize - 1 Step 1
                For nIdxWidth = 0 To nWidthSize - 1 Step 1
                    If (_token.IsCancellationRequested) Then
                        bRst = False
                        Exit For
                    End If

                    Dim pAdr As IntPtr = WriteableBitmap.BackBuffer
                    Dim nPos As Integer = nIdxHeight * WriteableBitmap.BackBufferStride + nIdxWidth * 4
                    Dim bytePixelB As Integer = ReadByte(pAdr, nPos + ComInfo.Pixel.B)
                    Dim bytePixelG As Integer = ReadByte(pAdr, nPos + ComInfo.Pixel.G)
                    Dim bytePixelR As Integer = ReadByte(pAdr, nPos + ComInfo.Pixel.R)

                    Dim byteGrayScale As Integer = (bytePixelB + bytePixelG + bytePixelR) / 3
                    WriteByte(pAdr, nPos + ComInfo.Pixel.B, CByte(byteGrayScale))
                    WriteByte(pAdr, nPos + ComInfo.Pixel.G, CByte(byteGrayScale))
                    WriteByte(pAdr, nPos + ComInfo.Pixel.R, CByte(byteGrayScale))
                Next
            Next
            WriteableBitmap.AddDirtyRect(New Int32Rect(0, 0, nWidthSize, nHeightSize))
            WriteableBitmap.Unlock()
            WriteableBitmap.Freeze()

            Return bRst
        End Function
    End Class
End Namespace

File: CSharp/ImageProcessing/GrayScale.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Media.Imaging;

namespace ImageProcessing
{
    public class GrayScale : ComImgProc
    {
        public GrayScale(BitmapImage _bitmap) : base(_bitmap)
        {
        }

        ~GrayScale()
        {
        }

        public override bool GoImgProc(CancellationToken _token)
        {
            bool bRst = true;

            int nWidthSize = base.m_bitmap.PixelWidth;
            int nHeightSize = base.m_bitmap.PixelHeight;

            base.m_wBitmap = new WriteableBitmap(base.m_bitmap);
            base.m_wBitmap.Lock();

            int nIdxWidth;
            int nIdxHeight;

            unsafe
            {
                for (nIdxHeight = 0; nIdxHeight < nHeightSize; nIdxHeight++)
                {
                    for (nIdxWidth = 0; nIdxWidth < nWidthSize; nIdxWidth++)
                    {
                        if (_token.IsCancellationRequested)
                        {
                            bRst = false;
                            break;
                        }

                        byte* pPixel = (byte*)base.m_wBitmap.BackBuffer + nIdxHeight * base.m_wBitmap.BackBufferStride + nIdxWidth * 4;
                        byte byteGrayScale = (byte)((pPixel[(int)ComInfo.Pixel.B] + pPixel[(int)ComInfo.Pixel.G] + pPixel[(int)ComInfo.Pixel.R]) / 3);

                        pPixel[(int)ComInfo.Pixel.B] = byteGrayScale;
                        pPixel[(int)ComInfo.Pixel.G] = byteGrayScale;
                        pPixel[(int)ComInfo.Pixel.R] = byteGrayScale;
                    }
                }
                base.m_wBitmap.AddDirtyRect(new Int32Rect(0, 0, nWidthSize, nHeightSize));
                base.m_wBitmap.Unlock();
                base.m_wBitmap.Freeze();
            }

            return bRst;
        }
    }
}
